src/app/qgshandlebadlayers.cpp

```diff
@@ -509,3 +495,164 @@ int QgsHandleBadLayers::layerCount()
 {
   return mLayerList->rowCount();
 }
+
+QString QgsHandleBadLayers::findFile( const QString &fileName, const QString &basePath, int maxDepth, int driveMargin )
+{
+  int depth = 0;
+  QDir folder = QDir( QDir( basePath ).absolutePath() );
+  QString existingBase;
+  // find the nearest existing folder
+  while ( !folder.exists() && folder.absolutePath().count( '/' ) > driveMargin )
+  {
+    existingBase = folder.path();
+    if ( !folder.cdUp() )
+      folder = QFileInfo( existingBase ).absoluteDir(); // using fileinfo to move up one level
+    depth += 1;
+  }
+  if ( depth > maxDepth )
+    maxDepth = depth;
+  while ( depth <= maxDepth && folder.cdUp() && folder.absolutePath().count( '/' ) > driveMargin )
+  {
+    QDirIterator finder( folder.path(), QStringList() << fileName, QDir::Files, QDirIterator::Subdirectories );
+    if ( finder.hasNext() )
+      return ( finder.next() );
+    else
+      depth += 1;
+  }
+  return ( existingBase + '/' + fileName );
+}
+
+QString QgsHandleBadLayers::checkBasepath( const QString &layerId, const QString &newPath, const QString &fileName )
+{
+  const QString originalBase = mOriginalFileBase.value( layerId );
+  const QFileInfo newpathInfo = QFileInfo( newPath );
+  if ( newpathInfo.exists() && newpathInfo.isFile() )
```

Comment:
Can you explain the logic regarding mAlternativeBasepaths for me?

Reply:
Ah yes, the logic is based on my first PR about the handler. I am assuming that if someone is moving folders entirely then when many of those layer break the changes that occured for one layer should be the same for every other layer.

This way I aimed to make a dictionary of viable paths for each basepath. IE if something in a given folder is split in 3 distinct folders, than the original path will be associated with the three new viable paths that are encountered and those are reused to speed up fetching.